python/cudf/cudf/core/column/column.py

```diff
@@ -97,6 +109,423 @@ class ColumnBase(Column, Serializable, BinaryOperand, Reducible):
+        and compute new data Buffers zero-copy that use pointer arithmetic to
+        properly adjust the pointer.
+        """
+        mask_size = plc.null_mask.bitmask_allocation_size_bytes(self.size)
+        required_num_bytes = -(-self.size // 8)  # ceiling divide
+        error_msg = (
+            "The value for mask is smaller than expected, got {} bytes, "
+            f"expected {required_num_bytes} bytes."
+        )
+        if value is None:
+            mask = None
+        elif hasattr(value, "__cuda_array_interface__"):
+            if value.__cuda_array_interface__["typestr"] not in ("|i1", "|u1"):
+                if isinstance(value, ColumnBase):
+                    value = value.data_array_view(mode="write")
+                value = cupy.asarray(value).view("|u1")
+            mask = as_buffer(value)
+            if mask.size < required_num_bytes:
+                raise ValueError(error_msg.format(str(value.size)))
+            if mask.size < mask_size:
+                dbuf = rmm.DeviceBuffer(size=mask_size)
+                dbuf.copy_from_device(value)
+                mask = as_buffer(dbuf)
+        elif hasattr(value, "__array_interface__"):
+            value = np.asarray(value).view("u1")[:mask_size]
+            if value.size < required_num_bytes:
+                raise ValueError(error_msg.format(str(value.size)))
+            dbuf = rmm.DeviceBuffer(size=mask_size)
+            dbuf.copy_from_host(value)
+            mask = as_buffer(dbuf)
+        else:
+            try:
+                value = memoryview(value)
+            except TypeError as err:
+                raise TypeError(
+                    f"Expected a Buffer object or None for mask, got {type(value).__name__}"
+                ) from err
+            else:
+                value = np.asarray(value).view("u1")[:mask_size]
+                if value.size < required_num_bytes:
+                    raise ValueError(error_msg.format(str(value.size)))
+                dbuf = rmm.DeviceBuffer(size=mask_size)
+                dbuf.copy_from_host(value)
+                mask = as_buffer(dbuf)
+
+        return cudf.core.column.build_column(  # type: ignore[return-value]
+            data=self.data,
+            dtype=self.dtype,
+            mask=mask,
+            size=self.size,
+            offset=0,
+            children=self.children,
+        )
+
+    @property
+    def null_count(self) -> int:
+        if self._null_count is None:
+            if not self.nullable or self.size == 0:
+                self._null_count = 0
+            else:
+                with acquire_spill_lock():
+                    self._null_count = plc.null_mask.null_count(
+                        self.base_mask.get_ptr(mode="read"),  # type: ignore[union-attr]
+                        self.offset,
+                        self.offset + self.size,
+                    )
+        return self._null_count
```

Comment:
*question*: pylibcudf columns claim to have a correct null count by construction, why must we launch a kernel here? Is it because we don't always start from a pylibcudf column?

Reply:
> Is it because we don't always start from a pylibcudf column?

Yes correct. At least in cudf classic there are routines that use `build_column` to make a new column, and `build_column` accepts a `null_count: int | None` so it appears cudf classic lazily computes the `null_count` if not provided.

Reply:
I think we should probably change this behavior in cudf classic. libcudf used to also support lazily determining the null count, but that was changed to support making libcudf stream-ordered (see https://github.com/rapidsai/cudf/issues/11968). Since pylibcudf follows suit automatically, we should probably do the same in cudf classic. No need to focus on that now since a lot of the work that would be necessary will change as we completely rewrite the column layer, but maybe open an issue to keep track of that.

Reply:
Yup agreed. I'd like the cudf classic column to mirror pylibcudf/libcudf columns as close as possible

Reply:
Opened https://github.com/rapidsai/cudf/issues/17816